### PR TITLE
Add feature for bonding/vlan interface in the firewall_interfaces

### PR DESCRIPTION
**Pull request by mlk-89**
Add vlan interface in the firewall_interfaces
Add bonding interface in the firewall_interfaces

##### SUMMARY
When using name that contains a dot in the module vyos_firewall_interfaces , we consider this interface as a vlan interface, so this PR include modification on the facts gathering and config file.

Moreover, set the interface type as well when the name begins with bond.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
plugins/module_utils/network/vyos/config/firewall_interfaces/firewall_interfaces.py
plugins/module_utils/network/vyos/facts/firewall_interfaces/firewall_interfaces.py

##### ADDITIONAL INFORMATION
Fix this issue: https://github.com/ansible-collections/vyos.vyos/issues/219

usage:   
vyos.vyos.vyos_firewall_interfaces:
    config:
    - access_rules:
      - afi: ipv4
        rules:
        - name: test1
          direction: in
        - name: test2
          direction: local
      name: eth1.10
    - access_rules:
      - afi: ipv4
        rules:
        - name: test1
          direction: in
        - name: test2
          direction: local
      name: bond0
    - access_rules:
      - afi: ipv4
        rules:
        - name: test1
          direction: in
        - name: test2
          direction: local
      name: bond0.113

### CODE REVIEW

**Comment by GomathiselviS** [comment]
@mlk-89  Thanks for working on this. Can you please add relevant test cases to test the fix and a change fragment to explain the fix briefly?

**Comment by mlk-89** [comment]
HI @GomathiselviS, yes sure I can write some tests, can you explain me how the testing environnment works?

Is there already a bonding interface that I can use? or should I create them from other tests?

**Comment by GomathiselviS** [comment]
@mlk-89 You can add a test case to https://github.com/ansible-collections/vyos.vyos/blob/main/tests/unit/modules/network/vyos/test_vyos_firewall_interfaces.py. The device connection is mocked in the unit tests. You can add a bonding interface config to https://github.com/ansible-collections/vyos.vyos/blob/main/tests/unit/modules/network/vyos/fixtures/vyos_firewall_interfaces_config.cfg and test. Please let me know if you have any queries.

**Comment by omnom62** [comment]
@mlk-89 hi! Are you able to fetch the latest changes from main branch into yours so I can review your PR, please? Thanks!

File: plugins/module_utils/network/vyos/facts/firewall_interfaces/firewall_interfaces.py [hunk 1]
```diff
@@ -88,10 +94,22 @@ def get_names(self, data, interfaces):
         """
         names = []
         for r in set(interfaces):
-            int_regex = r" %s .+$" % r.strip("'")
-            cfg = findall(int_regex, data, M)
-            fi = self.render_config(cfg)
-            fi["name"] = r.strip("'")
+            myvif = None
+            if type(r) is tuple:
```
Comment (omnom62) [inline plugins/module_utils/network/vyos/facts/firewall_interfaces/firewall_interfaces.py:98]: The recommended expression here is if isinstance(r, tuple):

**Review by omnom62 — CHANGES_REQUESTED** [review]
Hi @mlk-89 

Can you please fetch the latest changes to vyos.vyos into your fork?

I requested a change to use modern function to check for the type.
On the other note - I cannot see this code work for v.14+ as set interface <type> name firewall - is not a correct path (there is no firewall modifier in 1.4+)
Otherwise, I cloned the current vyos.vyos, made this and other syntax/formatting changes and successfully tested your code against v1.3.4, but not 1.4+

**Comment by mlk-89** [comment]
Hi @omnom62  Yes sure

**Comment by mlk-89** [comment]
There is a change on the command  'set firewall' for vyos 1.4

On 1.3: set firewall name
On 1.4: set firewall ipv4 name

Maybe we can add a check for this

**Comment by gaige** [comment]
The module itself likely doesn't need to be used for the 1.4 version, as @mlk-89 mentions. We already have support for the new filter rule types (input filter/output filter/forward filter) as well as the interface rules. When used with version 1.3, users will need to use the interface module, but for 1.4+ they'll need to switch up how they implement their firewalls to account for the new mechanism.

**Comment by gaige** [comment]
I enabled the checks for this PR, the sanity checks are showing a few small errors, can you update these?

```
ERROR: Found 8 pep8 issue(s) which need to be resolved:
ERROR: plugins/module_utils/network/vyos/config/firewall_interfaces/firewall_interfaces.py:397:9: E265: block comment should start with '# '
ERROR: plugins/module_utils/network/vyos/config/firewall_interfaces/firewall_interfaces.py:398:20: W291: trailing whitespace
ERROR: plugins/module_utils/network/vyos/config/firewall_interfaces/firewall_interfaces.py:400:24: W291: trailing whitespace
ERROR: plugins/module_utils/network/vyos/config/firewall_interfaces/firewall_interfaces.py:406:9: E265: block comment should start with '# '
ERROR: plugins/module_utils/network/vyos/config/firewall_interfaces/firewall_interfaces.py:413:48: E221: multiple spaces before operator
ERROR: plugins/module_utils/network/vyos/config/firewall_interfaces/firewall_interfaces.py:415:45: E221: multiple spaces before operator
ERROR: plugins/module_utils/network/vyos/facts/firewall_interfaces/firewall_interfaces.py:61:9: E265: block comment should start with '# '
ERROR: plugins/module_utils/network/vyos/facts/firewall_interfaces/firewall_interfaces.py:94:13: E265: block comment should start with '# '
```

A couple lines with trailing whitespace and some comments that start right after the `#`, which require the addition of a space.

**Comment by omnom62** [comment]
> I enabled the checks for this PR, the sanity checks are showing a few small errors, can you update these?
> 
> ```
> ERROR: Found 8 pep8 issue(s) which need to be resolved:
> ERROR: plugins/module_utils/network/vyos/config/firewall_interfaces/firewall_interfaces.py:397:9: E265: block comment should start with '# '
> ERROR: plugins/module_utils/network/vyos/config/firewall_interfaces/firewall_interfaces.py:398:20: W291: trailing whitespace
> ERROR: plugins/module_utils/network/vyos/config/firewall_interfaces/firewall_interfaces.py:400:24: W291: trailing whitespace
> ERROR: plugins/module_utils/network/vyos/config/firewall_interfaces/firewall_interfaces.py:406:9: E265: block comment should start with '# '
> ERROR: plugins/module_utils/network/vyos/config/firewall_interfaces/firewall_interfaces.py:413:48: E221: multiple spaces before operator
> ERROR: plugins/module_utils/network/vyos/config/firewall_interfaces/firewall_interfaces.py:415:45: E221: multiple spaces before operator
> ERROR: plugins/module_utils/network/vyos/facts/firewall_interfaces/firewall_interfaces.py:61:9: E265: block comment should start with '# '
> ERROR: plugins/module_utils/network/vyos/facts/firewall_interfaces/firewall_interfaces.py:94:13: E265: block comment should start with '# '
> ```
> 
> A couple lines with trailing whitespace and some comments that start right after the `#`, which require the addition of a space.

@gaige I actually fixed those in the two files I sent you - otherwise, the errors are easy to fix from scratch

**Comment by gaige** [comment]
> > The module itself likely doesn't need to be used for the 1.4 version, as @mlk-89 mentions. We already have support for the new filter rule types (input filter/output filter/forward filter) as well as the interface rules. When used with version 1.3, users will need to use the interface module, but for 1.4+ they'll need to switch up how they implement their firewalls to account for the new mechanism.
> 
> Perhaps we need to specify in the doc, that the module is deprecated in 1.4+, and this should do?

That would make sense. That'll be up in the `docs` directory. Probably just a simple note that this module is deprecated in 1.4+ and to incorporate directly into the `vyos.vyos_firewall_rules` for versions 1.4+

**Comment by gaige** [comment]
One more request is that we add tests. Something like this should work if placed in `tests/unit/modules/network/vyos/test_vyos_firewall_interfaces.py`:

```python
    def test_vyos_firewall_rule_set_02_replaced(self):
        set_module_args(
            dict(
                config=[
                    dict(
                        name="eth0.100",
                        access_rules=[
                            dict(
                                afi="ipv4",
                                rules=[dict(name="INBOUND", direction="in")],
                            ),
                            dict(
                                afi="ipv6",
                                rules=[dict(name="V6-LOCAL", direction="local")],
                            ),
                        ],
                    ),
                    dict(
                        name="bond2",
                        access_rules=[
                            dict(
                                afi="ipv4",
                                rules=[dict(name="LOCAL", direction="local")],
                            ),
                            dict(
                                afi="ipv6",
                                rules=[dict(name="V6-LOCAL", direction="local")],
                            ),
                        ],
                    ),
                    dict(
                        name="wg4",
                        access_rules=[
                            dict(
                                afi="ipv4",
                                rules=[dict(name="LOCAL", direction="local")],
                            ),
                            dict(
                                afi="ipv6",
                                rules=[dict(name="V6-LOCAL", direction="local")],
                            ),
                        ],
                    ),
                ],
                state="replaced",
            ),
        )
        commands = [
            'delete interfaces ethernet eth0 firewall in name',
            'delete interfaces ethernet eth0 firewall local name',
            'delete interfaces ethernet eth0 firewall out name',
            'delete interfaces ethernet eth0 firewall local ipv6-name',
            'delete interfaces ethernet eth2 firewall in name',
            'delete interfaces ethernet eth2 firewall local name',
            'delete interfaces ethernet eth2 firewall out name',
            'delete interfaces ethernet eth2 firewall local ipv6-name',
            "set interfaces ethernet eth0 vif 100 firewall in name 'INBOUND'",
            "set interfaces ethernet eth0 vif 100 firewall local ipv6-name 'V6-LOCAL'",
            "set interfaces bonding bond2 firewall local name 'LOCAL'",
            "set interfaces bonding bond2 firewall local ipv6-name 'V6-LOCAL'",
            "set interfaces wireguard wg4 firewall local name 'LOCAL'",
            "set interfaces wireguard wg4 firewall local ipv6-name 'V6-LOCAL'"
        ]
        self.execute_module(changed=True, commands=commands)
```

**Comment by gaige** [comment]
I raised a PR to your branch with the aforementioned changes for your convenience. Please review and see if they're appropraite.

**Review by gaige — APPROVED** [review]
LGTM. Thanks for your contribution

**Review by omnom62 — APPROVED** [review]
LGTM

**Review by sever-sever — APPROVED** [review]

**Pull request merged**